### ISSUE TITLE
`proxy-protocol = "v2"` doesn't work well

### ISSUE THREAD

**Issue by shawn0915** [pingcap/tiproxy#346]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

tiproxy --config tiproxy.toml

[proxy]
require-backend-tls = false
proxy-protocol = "v2"

### 3. What did you see instead (Required)

mysql --comments --host 127.0.0.1 --port 6000 -u root
ERROR 2013 (HY000): Lost connection to MySQL server at 'reading authorization packet', system error: 0

### 4. What is your version? (Required)

tiproxy: main,
tidb: 7.0.0

**Comment by djshow832** [comment]
Did you turn on the TiDB proxy protocol?

**Comment by shawn0915** [comment]
> Did you turn on the TiDB proxy protocol? https://docs.pingcap.com/tidb/stable/command-line-flags-for-tidb-configuration#--proxy-protocol-networks
> 
> > If this flag is left empty, no IP address can connect to TiDB using the PROXY protocol.

oh, ur right.
and, do you know how that show port `:0`?

```
mysql> \r
Connection id:    100
Current database: *** NONE ***

mysql> show processlist;
+--------------------+------+-------------+------+---------+------+------------+------------------+
| Id                 | User | Host        | db   | Command | Time | State      | Info             |
+--------------------+------+-------------+------+---------+------+------------+------------------+
| 151132271284322879 | root | 127.0.0.1:0 | NULL | Query   |    0 | autocommit | show processlist |
+--------------------+------+-------------+------+---------+------+------------+------------------+
1 row in set (0.01 sec)

```

**Comment by djshow832** [comment]
> > Did you turn on the TiDB proxy protocol? https://docs.pingcap.com/tidb/stable/command-line-flags-for-tidb-configuration#--proxy-protocol-networks
> > > If this flag is left empty, no IP address can connect to TiDB using the PROXY protocol.
> 
> oh, ur right. and, do you know how that show port `:0`?
> 
> ```
> mysql> \r
> Connection id:    100
> Current database: *** NONE ***
> 
> mysql> show processlist;
> +--------------------+------+-------------+------+---------+------+------------+------------------+
> | Id                 | User | Host        | db   | Command | Time | State      | Info             |
> +--------------------+------+-------------+------+---------+------+------------+------------------+
> | 151132271284322879 | root | 127.0.0.1:0 | NULL | Query   |    0 | autocommit | show processlist |
> +--------------------+------+-------------+------+---------+------+------------+------------------+
> 1 row in set (0.01 sec)
> ```

Good point. I didn't realize that yet. We'll investigate it and keep this issue. cc @xhebox

**Comment by xhebox** [comment]
Bug of tidb or go-proxyprotocol.

**Comment by YangKeao** [comment]
I think this issue is caused by the mix-use of IPv4 and IPv6 (especially for IPv4 mapped IPv6).

The Proxy Protocol always assumed that the family of the source and dest is the same, but the TiProxy doesn't follow the implementation:

```
	case *net.TCPAddr:
		addressFamily = ProxyAFINet
		if len(sadd.IP) == net.IPv6len {
			addressFamily = ProxyAFINet6
		}
		network = ProxyNetworkStream
		dadd, ok := dstAddr.(*net.TCPAddr)
		if !ok {
			return nil, ErrAddressFamilyMismatch
		}
		buf = append(buf, sadd.IP...)
		buf = append(buf, dadd.IP...)
		buf = append(buf, byte(sadd.Port>>8), byte(sadd.Port))
		buf = append(buf, byte(dadd.Port>>8), byte(dadd.Port))
```

If the `dstAddr` is IPv4, but `sadd` is IPv6, it'll produce wrong info. The bug in this issue is caused by the following ppv2 content:

```
[13 10 13 10 0 13 10 81 85 73 84 10 33 0 0 0 0 0 0 0 0 0 0 0 0 0 255 255 127 0 0 1 127 0 0 1 146 164 15 160]
```

From `::ffff:127.0.0.1:37540` to `127.0.0.1:4000`.
